pkg/controller/operator.go

```diff
@@ -2,77 +2,201 @@ package controller
+
+	// Send cancellation requests to both nodes
+	cancelReq := &pb.CancelMigrationRequest{
+		FromClusterID: op.fromShard.ShardID,
+		ToClusterID:   op.toShard.ShardID,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	// Cancel on source node
+	_, err = fromClient.CancelMigration(ctx, cancelReq)
+	if err != nil {
+		log.Printf("Error cancelling migration on source node: %v", err)
+		// Continue to cancel on destination even if source fails
+	}
+
+	// Cancel on destination node
+	_, err = toClient.CancelMigration(ctx, cancelReq)
+	if err != nil {
+		return fmt.Errorf("error cancelling migration on destination node: %v", err)
+	}
+
+	log.Printf("Migration cancelled between shards %d and %d", op.fromShard.ShardID, op.toShard.ShardID)
+	return nil
 }
 
-func (op *Operator) migrate(fromShard schema.RaftGroup, toShard schema.RaftGroup, keyranges []schema.KeyRange) error {
-	// Implement data migration logic here
-	log.Println("Migrating data from shard", fromShard.ShardID, "to shard", toShard.ShardID)
-	// fromIdx, toIdx := uint64(rand.Intn(len(fromShard.Members))), uint64(rand.Intn(len(toShard.Members)))
-	// fromNode, toNode := fromShard.Members[fromIdx], toShard.Members[toIdx]
-
-	// // Connect to source and destination nodes via gRPC
-	// fromAddr := fromNode.Address
-	// fromConn, err := grpc.NewClient(
-	// 	fmt.Sprintf("localhost:%d", 51000+fromNode.ID),
-	// 	grpc.WithTransportCredentials(insecure.NewCredentials()),
-	// )
-	// if err != nil {
-	// 	return fmt.Errorf("failed to connect to source node: %v", err)
-	// }
-	// defer fromConn.Close()
-	// fromClient := pb.NewNodeRouterClient(fromConn)
-
-	// toAddr := toNode.Address
-	// toConn, err := grpc.NewClient(
-	// 	fmt.Sprintf("localhost:%d", 51000+toNode.ID),
-	// 	grpc.WithTransportCredentials(insecure.NewCredentials()),
-	// )
-	// if err != nil {
-	// 	return fmt.Errorf("failed to connect to destination node: %v", err)
-	// }
-	// defer toConn.Close()
-	// toClient := pb.NewNodeRouterClient(toConn)
-
-	// // For each key range to migrate:
-	// for _, kr := range keyranges {
-	// 	log.Printf("Migrating range %v-%v from shard %d to shard %d",
-	// 		kr.Start, kr.End, fromShard.ShardID, toShard.ShardID)
-
-	// 	// 1. Read all keys in range from source
-	// 	readReq := &pb.ReadRequest{
-	// 		ClusterID: fromShard.ShardID,
-	// 		Key:       kr.Start, // Need to implement range reads
-	// 	}
-	// 	readResp, err := fromClient.Read(context.Background(), readReq)
-	// 	if err != nil {
-	// 		return fmt.Errorf("failed to read from source: %v", err)
-	// 	}
-
-	// 2. Write to destination
-	// writeReq := &pb.WriteRequest{
-	// 	ClusterID: toShard.ShardID,
-	// 	Key:       readResp.Key,
-	// 	Value:     readResp.Value,
-	// }
-	// _, err = toClient.Write(context.Background(), writeReq)
-	// if err != nil {
-	// 	return fmt.Errorf("failed to write to destination: %v", err)
-	// }
-	// }
+func (op *Operator) finishMigration() error {
```

Comment:
Should we notify only the source and destination nodes (two nodes) or all nodes?

Btw please consider adding more comments regarding the migration workflow, thx :)

Reply:
1. RN we only notify 2 nodes and ask them to propose `apply_schedule` to raft. Logic can be added that if these 2 nodes fail to do so, controller asks other nodes to do so.

2. Will draw a diagram about the migration workflow.